### DISCUSSION TITLE
PXE over WiFi

### DISCUSSION THREAD

**Discussion by wheelnut02** [ipxe/ipxe#354] [Q&A]
Help! I am probably doing something wrong (it would not be the first time). I would like to experiment with PXE booting over wireless. I can get it to work fine without any authentication, but as soon as I try to make it connect using WPA2, it fails to work. The errors suggest that the password is wrong. I have tried this with both a Samsung and an Asus netbooks using both a FritzBox and Zyxel wireless routers with no success when it comes to WPA2 and PSK.
I end up with [RXE: 22 x "Packet decryption error (http://ipxe.org/1c1f6602)"]

Any ideas as to what I could be doing wrong ?
Many thanks,
Martin

**Comment by NiKiZe** [comment]
Wireless parts of iPXE has not really been updated, the most secure variant that iPXE does support is WEP.

The better solution these days is Android or iOS phone with USB tethering and using that as a bridge to connect to the net. you still need iPXE on USB and can there include a script that does the bootstrapping to connect to the correct server. (Since DHCP will be provided by the phone bridge)

**Reply by wheelnut02** [reply]
That is a pity. I had read somewhere that wireless pxe had not had much work done on it lately, but it was an old post, so thought perhaps something had been done more recently. It is a shame really. I will have a go with a mobile phone and usb tethering. Fortunately, I do have a phone with a usb port in it - soon, I think that many modern phones may no longer have this option.

**Reply by NiKiZe** [reply]
I have not seen any Android or iOS device that does not have a data phone. Do you have an example of a device missing it?

Feel free to find anyone that is willing to provide time or other resources to do the needed development. But it would probably mean that drivers is also needed.

**Reply by wheelnut02** [reply]
I only mentioned the phone usb issue because I have heard that some phones (I think Apple were considering this at some point) would not have any ports in them, no ear phone and presumably no USB (unless the wireless charger had that facility).
Networking is an area that I do struggle with and I have managed to get PXE on a wired network working perfectly, but I am not sure how to persuade iPXE to recognise that there is an Android phone connected to it via USB and to then subsequently, treat it like a network card. Once I can get that to work, I should be ok for the rest of it.
I would like to thank you all for taking the time and trouble to answer my question - just did not realise that it was a lot more complex than I had originally throught.
If anyone can provide a little more guidance on how to get iPXE to recognise the mobile telephone via USB, I would be very grateful. This is not a work related thing, just something I like to learn about.
Thanks again.

**Reply by mcb30** [reply]
@wheelnut02 For phone tethering, you probably want a combination of the ECM, ACM (RNDIS) and iPhone drivers, which would be the `ecm--acm--iphone` target list.  For example, if you wanted to build a version of iPXE to load from a USB key on a BIOS system, you'd build as `make bin/ecm--acm--iphone.usb`.

**Reply by wheelnut02** [reply]
Got the phone tethering to work, or at least I think I have. Seem to be stumbling at every turn. I am using an android phone, so compiling using:
make bin/ecm--acm--cdc.usb
(it is a cdc-ether device).
The netbook "sees" the device in so much as it recognises that it is fastened to the USB bus. I can set an ip address manually, but the phone cannot ping it and it cannot ping the phone. 
Running ifstat net0 returns a mac address (which varies every time I reboot the netbook) using rf-rndis on a usb bus reference (Ethernet) [open] [Link:up, TX:9 TXE:0 RX:0 RXE:0]
dhcp net0 times out (not really that surprised about that though)
I must be missing out things here, sorry...

**Reply by NiKiZe** [reply]
Remember that the phone provides DHCP which is not the same as the network which the phone is connected to.

**Reply by wheelnut02** [reply]
Ahhhhhhhhhhhhhhhhhhhhhhhhhhh - I had not thought of that, so very sorry.

**Reply by wheelnut02** [reply]
I think that I need to step away from this for a while. I have set up an ip address on the netbook in the same range as the phone and the netbook cannot ping the phone, nor can the phone ping the netbook.
dhcp net0 returns:
Configuring (net0 <mac address>)..................... No configuration methods succeeded (http://ipxe.org/040ee119)
I know that the dhcp server is working when I reboot the netbook in to linux and plug in the phone - I get an IP address (which I should have examined more closely earlier to see that it was on a different subnet)
When I try:
ifconf -c dhcp net0, I get:
Configuring [dhcp] (net0 <mac address>)..................... Connection timed out (http://ipxe.org/4c106035)
That link suggests running a tcpdump on the dhcp server... unfortunately, this did not present anything useful.

**Reply by NiKiZe** [reply]
Why do you set an IP? You need to use DHCP from the phone.

**Reply by wheelnut02** [reply]
I set up the static address on the netbook because dhcp was not returning an ip address... thought perhaps that might work... it didn't :(
Keep thinking that I have missed a step somewhere or perhaps need to compile in more driver detail into the ipxe. I will keep trying.

**Reply by wheelnut02** [reply]
Having re-compiled with debug on dhcp, ipxe is sending a DHCPDISCOVER to the phone, but there is no DHCPOFFER. I have tested the phone with the netbook running linux and I know that the DHCP server on the phone is working fine as I get an IP address back. I have also tried with the company iPhone, but that has iOS 14 on it and oes not appear to be working at all.

**Reply by mcb30** [reply]
Have you enabled tethering on the phone?  Some Android devices may present as a non-functional Ethernet USB device until tethering is enabled via the phone's UI.

On the iPhone, you should get an on-screen prompt to trust the device.  iPXE will display "Waiting for link-up" until you confirm the trust on the phone.

**Reply by wheelnut02** [reply]
Yep - USB tethering is enabled on the phone. I have compiled iPXE with acm--ecm--ncm--cdc and have dhcp on DEBUG:3. It appears to detect the phone as I get a reference as follows:
ifstat net0
net0: <mac address> using rf-rndis on 0000:00:1d.7-7-1.0 [Ethernet] (open)
[Link:up, TX:8 TXE:0 RX:0 RXE:0]
on entering dhcp net0, iPXE goes into the DHCPDISCOVER phase and tries a number of times before erroring with:
......... No configuration methods succeeded (http://ipxe.org/040ee119).
I know that the DHCP server is running on the phone because it works when I connect it to linux - I can get an IP address from the phones' internal dhcp server (192.168.42.x).
What I cannot determine for sure is whether the DHCPDISCOVER packets are not getting out of the netbook, not being understood by the phone or the phone is simply not responding with DHCPOFFER packets.
I am convinced I am doing something wrong somewhere, but I am just not sure where. The phone that I am using is not anything fancy, it is an Archos something or other, but from tests using it as a USB network in linux suggests that it is functioning correctly.

**Reply by wheelnut02** [reply]
Finally managed to get hold of a proper android phone. Managed to make use of a Samsung Galaxy A21S and iPXE worked perfectly. Now, I am wondering if it is the version of android that is on my Archos phone or whether the phone simply does not work properly.
Please accept my thanks for all of your time, help and considerable patience. For my next trick, I will see if it is possible to upgrade this phone or whether it will be a lost cause.

**Reply by mcb30** [reply]
Glad you managed to get something working! :slightly_smiling_face: 

Do the two phones show up identically within the `dmesg` output under Linux?

**Reply by wheelnut02** [reply]
They are all very similar. All show up as a saying that they use the cdc_ether driver (rndis), the samsung device manufacturer is Samsung, the bush / Archos mobile phone and one of the tablets manufacturer is mediatek and the neocore tablet is neocore.
 
Nothing really stands out, I have attached the messages file to this chat (sanitised to remove identification, but not touching anything useful I hope, just name and domain of the netbook device) perhaps that may yield something useful to the development staff.

The Bush / Archos phone and the Acer tablet are both running Android 7.0. The Neocore tablet is Android 9.0, but I cannot recall what the Samsung phone is running - it is not immediately to hand so I would need to get details of that separately.

It is interesting that the iPXE software would manage to get a link up response suggesting that the connection was working to the Bush / Archos and tablet devices, but not receiving a DHCPOFFER packet back. The Samsung telephone responded instantly with a DHCPOFFER, followed instantly with a DHCPREQUEST, DHCPACK and finally an IP address.

I can only guess at this stage that perhaps there is something missing in the implementation on the Bush / Archos, Acer and Neocore versions of Android (could be also why those devices were so cheap!).

Hope that helps,

Martin
[messages.txt](https://github.com/ipxe/ipxe/files/6456913/messages.txt)

**Reply by mcb30** [reply]
Thanks.  There's nothing immediately obvious in the kernel logs to indicate a difference.  If you can identify a cheap (sub-$50) device available in the UK that will exhibit the same problem, I can order one and take a look.

**Reply by wheelnut02** [reply]
Hmmm - I have the Bush Spiro mobile phone which was about $120 (USD) at the time, both tablet devices were around the $150 (USD). I think pretty much any cheap "smart" phone that has wireless and usb tethering ability would exhibit the same behaviour. The Samsung Galaxy A21S was about $180 - that belongs to my Mother. Tempted to go to one of the second-hand computer stores to see if there is something there for a low price. Might end up with a large amount of junk though and I already have plenty of that.

**Reply by mcb30** [reply]
Would https://www.ebay.co.uk/itm/273977129048 be suitable for testing?

**Reply by wheelnut02** [reply]
It certainly looks similar to mine. My phone is a Bush Spira D5 Dual Camera running Android 7.0. The Android security patch level is 5 May 2017 (probably not relevant). Kernel version is 3.18.35. I have just had a look on ebay for anything more similar and I think yours is the closest - I just hope that it does the same thing mine does.
Is there anything I can try out on my device or setup that would help in the debugging ? It would be interesting to know why these android devices do not work in the same way that the samsung devices do.

**Reply by mcb30** [reply]
> It certainly looks similar to mine.

Thanks.  Have ordered it: should arrive by Monday for testing.

**Reply by mcb30** [reply]
Tested with the Bush/Archos phone from eBay: it worked first time with no problems:
```
iPXE 1.21.1+ (gd7bc9) -- Open Source Network Boot Firmware -- http://ipxe.org
Features: DNS HTTP iSCSI TFTP AoE ELF MBOOT PXE bzImage Menu PXEXT
net0: 36:08:90:e9:0b:3b using rf-rndis on 0000:00:04.0-5-1.0 (Ethernet) [closed]
  [Link:up, TX:0 TXE:0 RX:0 RXE:0]
iPXE> dhcp net0
Configuring (net0 36:08:90:e9:0b:3b)..... ok
iPXE> ifstat
net0: 36:08:90:e9:0b:3b using rf-rndis on 0000:00:04.0-5-1.0 (Ethernet) [open]
  [Link:up, TX:4 TXE:0 RX:9 RXE:3]
  [RXE: 2 x "The socket is not connected (http://ipxe.org/380a6001)"]
  [RXE: 1 x "Network unreachable (http://ipxe.org/28056011)"]
iPXE> route
net0: 192.168.42.41/255.255.255.0 gw 192.168.42.129
iPXE> kernel http://boot.ipxe.org/demo/boot.php  
http://boot.ipxe.org/demo/boot.php... ok
iPXE> imgstat
boot.php : 127 bytes [script] [SELECTED]
iPXE>
```
This is a "Bush Spira E2X 5" Smartphone" running Android 5.1 (kernel 3.10.65), attached to a WiFi network and exposed to iPXE via USB with tethering enabled.

**Reply by wheelnut02** [reply]
I think we have come full circle back to the original question where I said that I must be doing something wrong. May I ask what compile options you used so that I can try it out please ? Not just for me, but possibly for others that are struggling. Many thanks in advance.

**Reply by mcb30** [reply]
I'm just using `make bin/ecm--acm--iphone.usb` on the unmodified current master branch with no custom options at all.

**Reply by wheelnut02** [reply]
I think perhaps that was where I might have been going wrong. I was using make bin-x86_64-bios/acm-ecm-ncm.usb - i.e. not using the iphone option (figured since the phone was android, iphone would be inappropriate). Once my lab (which is in bits at the moment - moving from one place to another) is re-assembled, I will try it out on my android phone.
Thank you so very much for taking the time, trouble, patience and money to investigate this for me - I think I owe you £30 + beer.

**Reply by mcb30** [reply]
The `iphone` part should indeed be irrelevant, since there is no hardware plugged in that would use that driver.

I've quickly tested also with `bin/acm--ecm--ncm.usb` (and the somewhat obscure 64-bit BIOS build `bin-x86_64-pcbios/acm--ecm--ncm.usb`) and both work with no problems.

**Reply by wheelnut02** [reply]
Looks like I have opened a can of worms here. Unfortunately, I must have something wrong either with what I am doing and / or that there is something wrong with the telephone and / or there is something wrong with the netbooks - it still does not work. It worked fine with the Samsung Galaxy A21S - and given the success that mcb30 has had, makes me wonder what else I can try. Since the Bush Spiro phone works to present an IP address to linux when USB tethered... perhaps there is a problem with the USB elements of the netbooks - could be a BIOS thing... and yes, I am cluthing at straws.

**Reply by mcb30** [reply]
@wheelnut02 If you're comfortable with using VMs (e.g. via `virt-manager`), you could try passing the Bush Spiro phone through to a VM running `ipxe.usb`.  That would allow you to test the interaction of iPXE with your phone, independent of any potential issues with the netbooks.

**Reply by wheelnut02** [reply]
I have attempted to do this via VirtualBox and via an old Asus Desktop PC - no effect. I think that the ipxe is probably ok, but not convinced that the phone is working correctly. I have had this working on a Samsung A21S phone, so I know that ipxe works, but that was running Android 10. Everything that I own that is running android is either version 2 (HTC Wildfire), version 7 (Bush Spiro phone and Acer tablet) and version 9 (another tablet that I have). I am unable to confirm whether or not the phone is being recognised by the ipxe software - i.e. ipxe cannot "see" it because the phone is not saying that it is there or whether I need to compile in something specific to ipxe for these android phones to work. I have tried the make bin/ecm--acm--iphone.usb options that you have described, but it is almost like the phone needs something else enabling. The tethering is on and cable connected... I will try to obtain an android 10 device to confirm that the build of ipxe that I am using will work to eliminate ipxe as the cause. I do not think that the netbooks are causing any problems, but at this stage, just about anything is possible.

**Reply by mcb30** [reply]
If the `ifstat` command shows an `rf-rndis` network interface, then that would mean that iPXE is able to see the phone (i.e. that all of the USB passthrough configuration is correct)

**Reply by wheelnut02** [reply]
That is confirmed as working. I have the phone plugged in, USB tethering enabled and I get the rf-rndis reference. if I turn off the USB tethering, and then run ifstat, the rf-rndis disappears. When the rf-rndis reference is visible and I run dhcp net0 (I have dhcp debug set to 3), I get a couple of DHCPDISCOVER messages and then it fails with No configuration methods succeeded (http://ipxe.org/040ee119)... this would suggest to me that the DHCP server on the phone was not responding to the DHCPDISCOVER packets - I have no clue what I can do about that unfortunately. I have tried this on my android 2, other android 7 and an android 9 device all with the same effect. I have asked a colleague to tried this out on his android 9 device - hopefully, his will work. I know it works on android 10. Perhaps there is a setting on the android device other than USB tethering being enabled that may need to be changed, but I have no idea what that would be. I have even been in to "developer mode" to see if there was anything obvious, but unfortunately not :(

**Reply by wheelnut02** [reply]
Just tried this on my KVM configuration - unfortunately, it appears that it is still not happy. Same issue as above where I get "No configuration methods succeeded (http://ipxe.org/040ee119)"... :(

**Reply by wheelnut02** [reply]
Incidentally, just noticed that it returns 'RXE: 1 x "Operation not supported (http://ipxe.org/3c086003)"' error - not sure if that helps here...

**Reply by wheelnut02** [reply]
![image](https://user-images.githubusercontent.com/83869073/119622520-23ced600-bdff-11eb-8b45-79e9494f473f.png)
Above are the sort of errors that are being produced when I try to make this work. The command that I am using to compile the ipxe software is:

make bin/acm--ecm--ncm--cdc--usb.usb DEBUG=dhcp:3,netdevice:3,ifmgmt:3 EMBED=./shell.ipxe

**Reply by mcb30** [reply]
> Incidentally, just noticed that it returns 'RXE: 1 x "Operation not supported (http://ipxe.org/3c086003)"' error - not sure if that helps here...

That's actually a positive result - it proves that the receive datapath is working since you were able to receive an IPv6 packet (which was not recognised by your iPXE build, which does not have IPv6 enabled in the default configuration).

From the screenshot, everything looks fine - transmissions are completing and some packets are received.  The only problem is that you never get a response to the DHCPDISCOVER.

I'm out of things to suggest at this point, sorry, unless you feel like shipping one of the affected devices to Cambridge (UK).

**Reply by wheelnut02** [reply]
Certainly very tempted to send one of the devices to Cambridge (I am in Cheshire). I will see if I get exactly the same messages from the other 3 android devices. Curious as to why it was making use of IPv6 - is there a way to enable it so that there is a chance that the DHCP server on the phone will work ? Not sure whether the phone is capable of IPv6, but perhaps it is. At the end of the day, if I send one of my duff devices down to you, and you are able to diagnose a reason for it not behaving itself, it might help others who are interested in this problem.

**Reply by mcb30** [reply]
@wheelnut02 Email me directly (`mcb30@ipxe.org`) if you want to arrange for posting a device to me (and getting it back afterwards!)

**Reply by mcb30** [reply]
@wheelnut02 Good news: I am able to reproduce the problem using the Acer tablet that you sent to me!

**Reply by wheelnut02** [reply]
Thanks for that @mcb30  - at least it was not something that I was doing incorrectly. I would be very interested in knowing why it does not work. Not sure how complicated that would be to determine though.

**Reply by mcb30** [reply]
Continuing this discussion as #377

**Comment by sebaxakerhtc** [comment]
Ok, guys...
Theme started about iPXE over wifi. Not about android or iOS...
How can we help about working WPA2 connections?

**Reply by NiKiZe** [reply]
That question was answered, someone needs to write wpa2 code for iPXE, maybe it is better we just remove the existing WiFi code instead.

**Reply by sebaxakerhtc** [reply]
> maybe it is better we just remove the existing WiFi code instead.

May be... When it difficult to resolve - many people just skip it, when won't to resolve ;)

**Comment by sebaxakerhtc** [comment]
So... I found latest build with working WPA2 connection – [THIS build](https://github.com/ipxe/ipxe/tree/0fb37a48a9405a6b3b0c5b198bdd87f1a4bb51f0)
And from [THIS point](https://github.com/ipxe/ipxe/commit/860620459589444fd3c7c329aa7b177b5c9a4039) we have WPA2 4-way handshake timeout
@wheelnut02  Try this :)

**Reply by AdamKearn** [reply]
Hi @sebaxakerhtc,

I am trying to get wifi iPXE booting working as we have about 30+ laptops that stupidly don't come with a network port.
I've been able to compile iPXE with AXGE (aka ASIX) drivers which allows me to use a USB 2 Ethernet dongle.

I've tried to download that build you linked above and compile it but I just run into these errors:
Can you share what you did to get a successful build working?

Regards.

![image](https://user-images.githubusercontent.com/30593259/212089034-673939dd-23ae-493e-a0fc-b68ac1efa921.png)

**Reply by brybalicious** [reply]
Did it work?

**Comment by brybalicious** [comment]
What is the current status of PXE boot over WiFi?

Many Single Board Computers (SBC)'s e.g. Raspberry Pi Zero... could seriously benefit from PXE over WiFi, since the lack of requirement for an SD card means they avoid file corruption when the boards inevitably suffer incomplete shutdowns. They don't have robust power electronics nor resilience against sudden shutdown built in. Thus PXE boot solves a lot of their problems in a surprising way.

**Reply by NiKiZe** [reply]
As stated here, and in https://github.com/ipxe/ipxe/discussions/851#discussioncomment-4672336 WiFi support isn't really viable as native iPXE drivers.

For the Pi Zero it uses chips which requires a binary firmware to be uploaded to the NIC to be usable, it is unlikely that iPXE will ever want to include that. But there is plenty of device firmware themselves that has the drivers for the specific nic, and a way to configure and setup the NIC (ssid and psk). So if the Pis own firmware would include the drivers and an interface to connect to a network, and expose that as a EFI standard SNP/NII device, then iPXE can use that. There might already be drivers for this that can be loaded as EFI applications, but remember it is specific drivers for every wifi card out there, but iPXE can then use that driver, Have you tried to reach out to Broadcom, or the Pi Foundation for such a driver?

**Reply by brybalicious** [reply]
Thank you for the fast response. In fact, I'm trying to PXE boot an Orange Pi Zero 3W which uses a CDTech 20800D8-00 NIC. I've reached out to Orange Pi but no response yet. I'll continue to see if I can make progress there.

**Reply by brybalicious** [reply]
I can see from Taobao that it's synonymous with an [AIC8800D80](https://img.alicdn.com/imgextra/i4/2219841631759/O1CN01HrmWoy1OrfkzxSy5z_!!2219841631759.jpg_300x300q50.jpg_.webp) SoC which upon further digging is a Realtek [RTL8800](https://www.aliexpress.com/p/wiki/article.html?keywords=aic8800d80-usb-wifi-driver-window-10). Does that mean Realtek drivers would work for this chip? Am I looking for a .bin file?

**Reply by brybalicious** [reply]
Promising?
https://github.com/gtxaspec/aic8800-wifi/tree/master

https://deepwiki.com/gtxaspec/aic8800-wifi/5-hardware-interfaces

Problem is there are so many .bin files, how do you know which one to use?

**Reply by NiKiZe** [reply]
I will repeat what I said before, iPXE will not have any native drivers that requires a binary blob (firmware bin)
What is supported is a efi driver that exposes the NIC as a standard EFI NIC that iPXE can use, do NOT confuse or think that any kind of Linux driver can be used for this.

**Reply by brybalicious** [reply]
oh I see. So I need to ask the manufacturer if they have such an EFI based driver available. Thanks for clarifying.

**Reply by AdamKearn** [reply]
I spent a lot of time on this trying to get this working at my workplace.

I was trying to get this to integrate with FOG (our PXE deployment system)

I instead added the WiFi driver into the Linux kernel/image compiled this as a bootable image instead.

Then I just load this directly with iPXE and that then connects to the WiFi and starts imaging.

I would still prefer getting WiFi drivers directly within iPXE as it means you don't have to mess around with Linux but this is a good option for now

**Reply by NiKiZe** [reply]
@AdamKearn Out of curiosity, which brand of devices are you using?
Https boot over Wifi is supported by some vendors. (a few of them require Secure boot to be enabled for the option to exist)

**Reply by AdamKearn** [reply]
> @AdamKearn Out of curiosity, which brand of devices are you using?
> 
> Https boot over Wifi is supported by some vendors. (a few of them require Secure boot to be enabled for the option to exist)

HP probook.  These do have WiFi PXE in the BIOS on some of the G9 models that I've noticed.

I could not get this working in my testing.

I created a new SSID without PSK but it would still not remember the connection when rebooting into the "adapter"
